plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java

```diff
@@ -286,7 +318,40 @@ private int[] getWriterIndexes(Page page)
             }
 
             Optional<PartitionData> partitionData = getPartitionData(pagePartitioner.getColumns(), page, position);
-            writer = createWriter(partitionData);
+            // prepend query id to a file name so we can determine which files were written by which query. This is needed for opportunistic cleanup of extra files
+            // which may be present for successfully completing query in presence of failure recovery mechanisms.
+            String fileName = fileFormat.toIceberg().addExtension(session.getQueryId() + "-" + randomUUID());
+            String outputPath = partitionData
+                    .map(partition -> locationProvider.newDataLocation(partitionSpec, partition, fileName))
+                    .orElseGet(() -> locationProvider.newDataLocation(fileName));
+
+            ImmutableList.Builder<Integer> sortColumnIndexes = ImmutableList.builder();
+            ImmutableList.Builder<SortOrder> sortOrders = ImmutableList.builder();
+            for (TrinoSortField sortField : sortOrder) {
+                Types.NestedField column = outputSchema.findField(sortField.getSourceColumnId());
+                sortColumnIndexes.add(outputSchema.columns().indexOf(column));
```

Comment:
Could you pls verify whether `outputSchema.columns()` lookup still works in the context of dealing with nested types?

Reply:
Good catch, did not think about sorting on non-base columns.